pyclam/search.py

```diff
@@ -0,0 +1,196 @@
+        elif points.shape[1:] != self.data.shape[1:]:
+            raise ValueError(f'wrong dimensionality of points. Expected {self.data.shape[1:]}, but got {points.shape[1:]}')
+        else:
+            return
+
+    def _parse_query(self, query: Datum) -> np.ndarray:
+        if isinstance(query, int):
+            query: np.ndarray = np.asarray([self.data[query]])
+        query = np.expand_dims(query, 0)
+        self._check_shape(query)
+        return query
+
+    def tree_search(self, query: Datum, radius: float, start: Cluster, *, max_depth: Optional[int] = None) -> ClusterResults:
+        """ Performs tree-search for the query, starting at the given cluster.
+
+        :param query: point around which to search.
+        :param radius: distance from point within which to look.
+        :param start: starting cluster for search.
+        :param max_depth: optional maximum search depth starting at Cluster.
+        :return: dictionary of cluster -> distance between query and cluster center
+        """
+        if max_depth is None:
+            max_depth = self.manifold.depth
+        elif max_depth < 0:
+            raise ValueError(f'max_depth must be a non-negative integer. Got {max_depth} instead.')
+        else:
+            max_depth = min(self.manifold.depth, start.depth + max_depth)
+
+        query = self._parse_query(query)
+        candidates: ClusterResults = {start: -1}
+        hits: ClusterResults = dict()
+        for depth in range(start.depth, max_depth + 1):
+            if len(candidates) == 0:
+                break
+            # find distances to centers of candidate clusters
+            clusters: List[Cluster] = list(candidates.keys())
+            centers: np.ndarray = np.asarray([cluster.center for cluster in clusters])
+            self._check_shape(centers)
+            distances: List[float] = list(self.distance(query, centers)[0])
+
+            candidates = {  # filter out clusters that are too far away
+                cluster: distance for cluster, distance in zip(clusters, distances)
+                if distance <= (radius + cluster.radius)
+            }
+            subsumed: Set[Cluster] = {  # set of clusters that lie completely within radius of query
+                cluster for cluster, distance in candidates.items()
+                if (distance + cluster.radius) <= radius
+            }
+            hits.update({  # add subsumed clusters to hits, singletons should already be included here
+                cluster: distance
+                for cluster, distance in candidates.items()
+                if cluster in subsumed
+            })
+            hits.update({  # add non-singleton leaves to hits
+                cluster: distance
+                for cluster, distance in candidates.items()
+                if (cluster not in subsumed) and (len(cluster.children) == 0)
+            })
```

Comment:
Seems like these two comprehensions could be merged

Reply:
done